lib/icmp/icmpv4_print.ml

```diff
@@ -1,10 +1,58 @@
-let echo_request id seq =
-  let open Icmpv4_wire in
-  (* TODO: can this just be an appropriately-sized cstruct? *)
-  let buf = Io_page.(get 1 |> to_cstruct) in
-  let buf = Cstruct.set_len buf sizeof_icmpv4 in
-  set_icmpv4_ty buf 0x08;
-  set_icmpv4_code buf 0x00;
-  set_icmpv4_seq buf seq;
-  set_icmpv4_id buf id;
-  buf
+open Icmpv4_wire
+
+let echo ~buf ?payload ~ty ~id ~seq =
+  let min_len = sizeof_icmpv4 + (match payload with | None -> 0 | Some payload ->
+      Cstruct.len payload) in
+  if Cstruct.len buf < min_len then
+    Result.Error "Not enough space for ICMP header and payload"
+  else begin
+    let buf = Cstruct.set_len buf min_len in
+    set_icmpv4_ty buf (ty_to_int ty);
+    set_icmpv4_code buf 0x00;
+    set_icmpv4_csum buf 0x0000;
+    set_icmpv4_seq buf seq;
+    set_icmpv4_id buf id;
+    let packets = match payload with
+      | Some payload -> [buf ; payload]
+      | None -> [buf]
+    in
+    set_icmpv4_csum buf (Tcpip_checksum.ones_complement_list packets);
+    Result.Ok ()
+  end
+
+let echo_request ~buf ?payload ~id ~seq =
+  echo ~buf ?payload ~ty:Echo_request ~id ~seq
+
+let echo_reply ~buf ?payload ~id ~seq =
+  echo ~buf ?payload ~ty:Echo_reply ~id ~seq
+
+(** [would_fragment ip_header ip_payload next_hop_mtu] generates an
+    ICMP destination unreachable message, with the code set to 4 ("packet
+    fragmentation is required but the don't-fragment bit is set").  [ip_header] should
+    be the IP header of the packet which will be rejected; [ip_payload] is the
+    optional first 8 bytes of the IPv4 payload (generally the first 8 bytes of
+    a UDP or TCP header).  Payloads of length > 8 will be truncated. *)
+let would_fragment ~buf ~ip_header ?ip_payload ~next_hop_mtu =
```

Comment:
See https://github.com/docker/vpnkit/blob/master/src/hostnet/lib/slirp.ml#L74 for an example.  A proper NAT implementation would have to include this (among other ICMP pieces) as well.